sonic_data_client/db_client.go

```diff
@@ -925,97 +1068,111 @@ func dbTableKeySubscribe(gnmiPath *gnmipb.Path, c *DbClient) {
-		return
+	signalSync()
+
+	// Clear the payload so that next time it will send only updates
+	if updateOnly {
+		msiAll = make(map[string]interface{})
 	}
-	// First sync for this key is done
-	c.synced.Done()
-	for k := range msi {
-		delete(msi, k)
+
+	// Start routines to listen on the table changes.
+	updateChannel := make(chan map[string]interface{})
+	for _, rsd := range rsdList {
+		go dbSingleTableKeySubscribe(c, rsd, updateChannel)
 	}
+
+	// Listen on updates from tables.
+	// Depending on the interval, send the updates every interval or on change only.
+	intervalTicker := make(<-chan time.Time)
 	for {
+
+		// The interval ticker ticks only when the interval is non-zero.
+		// Otherwise (e.g. on-change mode) it would never tick.
+		if interval > 0 {
+			intervalTicker = IntervalTicker(interval)
+		}
+
 		select {
-		default:
-			val = nil
-			err = nil
-			c.mu.Lock()
-			if len(msi) > 0 {
-				val, err = msi2TypedValue(msi)
-				for k := range msi {
-					delete(msi, k)
+		case updatedTable := <-updateChannel:
+			log.V(1).Infof("update received: %v", updatedTable)
+			if interval == 0 {
+				// on-change mode, send the updated data.
+				if err := sendMsiData(updatedTable); err != nil {
+					handleFatalMsg(err.Error())
+					return
+				}
+			} else {
+				// Update the overall table, it will be sent when the interval ticks.
+				for k := range updatedTable {
+					msiAll[k] = updatedTable[k]
 				}
 			}
-			c.mu.Unlock()
-			if err != nil {
-				enqueFatalMsg(c, err.Error())
+		case <-intervalTicker:
+			log.V(1).Infof("ticker received: %v", len(msiAll))
+
+			if err := sendMsiData(msiAll); err != nil {
+				handleFatalMsg(err.Error())
 				return
 			}
-			if val != nil {
-				spbv = &spb.Value{
-					Path:      gnmiPath,
-					Timestamp: time.Now().UnixNano(),
-					Val:       val,
-				}
 
-				log.V(5).Infof("dbTableKeySubscribe enque: %v", spbv)
-				if err = c.q.Put(Value{spbv}); err != nil {
-					log.V(1).Infof("Queue error:  %v", err)
-					return
-				}
+			// Clear the payload so that next time it will send only updates
+			if updateOnly {
+				msiAll = make(map[string]interface{})
+				log.V(1).Infof("msiAll cleared: %v", len(msiAll))
 			}
 
-			// check possible value change every 100 millisecond
-			// TODO: make all the instances of wait timer consistent
-			time.Sleep(time.Millisecond * 100)
 		case <-c.channel:
 			log.V(1).Infof("Stopping dbTableKeySubscribe routine for %v ", c.pathG2S)
 			return
 		}
 	}
 }
 
-func  (c *DbClient) Set(path *gnmipb.Path, t *gnmipb.TypedValue, flagop int) error {
+func (c *DbClient) Set(path *gnmipb.Path, t *gnmipb.TypedValue, flagop int) error {
 	return nil
 }
-func (c *DbClient) Capabilities() ([]gnmipb.ModelData) {
+func (c *DbClient) Capabilities() []gnmipb.ModelData {
 	return nil
 }
 
+// validateSampleInterval validates the sampling interval of the given subscription.
+func validateSampleInterval(sub *gnmipb.Subscription) (time.Duration, error) {
```

Comment:
Do you know what is default behavior if someone passes garbage value like negative number as samplingInterval?

Reply:
It is not possible to pass a negative value because gRPC cannot accept a negative value as unsigned int. In other words, it is rejected from the middleware. E.g. this is what I got:
```
muacikgo@muacikgo-dev1:~/src/sonic-telemetry$ ./build/bin/gnmi_cli --client_types=gnmi -a 127.0.0.1:8080 -logtostderr -insecure -timestamp on -v 0 -dt p  --proto "$(cat updates-only-query.txt)"
F1209 22:49:41.509851  548707 gnmi_cli.go:269] failed to parse gNMI SubscribeRequest text proto: line 16: invalid uint64: -1
```

However, if someone passes a positive value that less than threshold, it is rejected with `invalid interval` message. E.g.
```
        client had error while displaying results:
        rpc error: code = InvalidArgument desc = invalid interval: 1ms. It cannot be less than 1s
```